cibyl/orchestrator.py

```diff
@@ -200,26 +202,49 @@ def run_features(self, system: System, features_to_run: list) -> None:
             system.populate(features_combination)
 
     def sort_and_filter_args(self) -> List[Argument]:
-        """Get a list of arguments that provide the queries that the sources
-        should perform. This requires filtering out the argument that have no
-        func associated, sorting the arguments by level and then making sure
-        that only one of the arguments is kept if many point to the same query
-        (e.g --builds and --build-status).
-        :returns: List of arguments that contain queries for the sources
+        """Select which arguments should be used to query the sources.
+
+        The algorithm used filters out the arguments with no func attribute,
+        then sorts the remaining ones by level. For each argument the shortest
+        path to a root node in the graph of queries is constructed and all
+        input arguments with func values contained in that path are
+        eliminated. When all input arguments have been consumed, the remaining
+        arguments in the queries list are the deepes nodes in each branch of
+        the queries graph, which corresponds to the queries that should be
+        made to the sources.
+
+        :returns: List of arguments that contain the queries to make to the
+        sources.
         """
-        args_with_query = {name: arg for name, arg in
-                           self.parser.ci_args.items() if arg.func}
-        sorted_args = sorted(args_with_query.values(),
+
+        args = [arg for arg in self.parser.ci_args.values() if arg.func]
+        sorted_args = sorted(args,
                              key=operator.attrgetter('level'), reverse=True)
-        queries = set()
-        final_args_to_query = []
-        for arg in sorted_args:
-            if arg.func not in queries:
-                queries.add(arg.func)
-                final_args_to_query.append(arg)
-        return final_args_to_query
+        nodes_visited = set()
+        graph = self.parser.graph_queries
+        roots = list(v for v, d in graph.in_degree() if d == 0)
+        queries = []
+        while sorted_args:
+            arg = sorted_args.pop(0)
+            if arg.func in nodes_visited:
+                # if the func was found in the path for a previous argument,
+                # skip it
+                continue
+            nodes_visited.add(arg.func)
+            queries.append(arg)
+            for root in roots:
+                # all the nodes in the path from the argument's func to a root
+                # node need not be queried
+                for path in nx.all_simple_paths(graph, source=root,
+                                                target=arg.func):
+                    # handle the case where there might be more than one path
+                    # between the node and the root, e.g. get_jobs has two
+                    # paths to the root (get_tenants)
+                    nodes_visited.update(path)
```

Comment:
I do not understand what 'nodes_visited' is supposed to store after this. Before you were storing function names but now it updates with a 'path', which I do not know what type it is or what it refers to here.

Reply:
Here `path` contains a list of nodes (function names) that are found between the root and the argument function. The `nodes_visited` stores all such nodes, so the arguments that are redundant can be skipped. A simple example would be a call like `cibyl --jobs job-name --last-build --tests`, the `path` variable for the `tests` argument would have the value `[get_builds, get_jobs]`, so both `last_build` and `jobs` will be ignored. That said I'll rename the `path` variable to something more explicit and expand the comments in this part to clarify